src/Microsoft.OpenApi.OData.Reader/Common/EdmModelHelper.cs

```diff
@@ -161,19 +161,25 @@ internal static IList<string> RetrieveNavigationPropertyPathsOperationIdSegments
                 navigationSource.Name
             };
 
-            IEnumerable<ODataNavigationPropertySegment> navPropSegments = path.Segments.Skip(1).OfType<ODataNavigationPropertySegment>();
-            Utils.CheckArgumentNull(navPropSegments, nameof(navPropSegments));
+            // For navigation property paths with odata type cast segments
+            // the OData type cast segments identifiers will be used in the operation id
+            IEnumerable<ODataSegment> segments = path.Segments.Skip(1).Where(s => s is ODataNavigationPropertySegment || s is ODataTypeCastSegment);
+            Utils.CheckArgumentNull(segments, nameof(segments));
 
-            foreach (var segment in navPropSegments)
+            string previousTypeCastSegmentId = null;
+            foreach (var segment in segments)
             {
-                if (segment == navPropSegments.Last())
+                if (segment is ODataNavigationPropertySegment navPropSegment)
                 {
-                    items.Add(segment.NavigationProperty.Name);
-                    break;
+                    items.Add(navPropSegment.NavigationProperty.Name);
                 }
-                else
+                else if (segment is ODataTypeCastSegment typeCastSegment && path.Kind == ODataPathKind.NavigationProperty)
                 {
-                    items.Add(segment.NavigationProperty.Name);
+                    // Only the last OData type cast segment identifier is added to the operation id
+                    items.Remove(previousTypeCastSegmentId);
```

Comment:
Should we check if `previousTypeCastSegmentId` is null before trying to remove. There looks to be a possible path if its the first time entering the block.

Reply:
There's no side-effect of `IEnumerable.Remove(null)`. No item in the collection will be removed, if there are any.